desc/optimizable.py

```diff
@@ -0,0 +1,155 @@
+            if not len(p):
+                warnings.warn(
+                    f"Object {self} was subclassed from Optimizable but no "
+                    + "optimizable parameters were declared"
+                )
+        return self._optimizable_params
+
+    @property
+    def params_dict(self):
+        """dict: dictionary of arrays of optimizable parameters."""
+        return {
+            key: jnp.asarray(getattr(self, key)).copy()
+            for key in self.optimizable_params
+        }
+
+    @params_dict.setter
+    def params_dict(self, d):
+        for key, val in d.items():
+            if jnp.asarray(val).size:
+                setattr(self, key, val)
+
+    @property
+    def dimensions(self):
+        """dict: dictionary of integers of sizes of each optimizable parameter."""
+        return {
+            key: jnp.asarray(getattr(self, key)).size for key in self.optimizable_params
+        }
+
+    @property
+    def x_idx(self):
+        """dict: dictionary of arrays of indices into array for each parameter."""
+        dimensions = self.dimensions
+        idx = {}
+        dim_x = 0
+        for arg in self.optimizable_params:
+            idx[arg] = jnp.arange(dim_x, dim_x + dimensions[arg])
+            dim_x += dimensions[arg]
+        return idx
+
+    @property
+    def dim_x(self):
+        """int: total number of optimizable parameters."""
+        return sum(
+            jnp.asarray(getattr(self, key)).size for key in self.optimizable_params
+        )
+
+    def pack_params(self, p):
+        """Convert a dictionary of parameters into a single array.
+
+        Parameters
+        ----------
+        p : dict
+            Dictionary of ndarray of optimizable parameters.
+
+        Returns
+        -------
+        x : ndarray
+            optimizable parameters concatenated into a single array, with indices
+            given by ``x_idx``
+        """
+        return jnp.concatenate(
+            [jnp.atleast_1d(p[key]) for key in self.optimizable_params]
+        )
+
+    def unpack_params(self, x):
+        """Convert a single array of concatenated parameters into a dictionary.
+
+        Parameters
+        ----------
+        x : ndarray
+            optimizable parameters concatenated into a single array, with indices
+            given by ``x_idx``
+
+        Returns
+        -------
+        p : dict
+            Dictionary of ndarray of optimizable parameters.
+        """
+        x_idx = self.x_idx
+        params = {}
+        for arg in self.optimizable_params:
+            params[arg] = jnp.atleast_1d(x[x_idx[arg]])
+        return params
+
+    def _sort_args(self, args):
+        """Put arguments in a canonical order. Returns unique sorted elements.
+
+        Actual order doesn't really matter as long as its consistent, though subclasses
+        may override this method to enforce a specific ordering
+        """
+        return sorted(set(list(args)))
```

Comment:
Does this order really not matter? Because `self.x_idx` depends on this order, so subclasses have to make sure they override that too.

Reply:
this is really only used once when optimizable_params is first created, after that everything uses the ordering of the optimizable_params list